src/robot/running/bodyrunner.py

```diff
@@ -334,21 +335,29 @@ def __init__(self, context, run=True, templated=False):
     def run(self, data):
         run = self._run
         executed_once = False
-        result = WhileResult(data.condition)
+        result = WhileResult(data.condition, data.limit)
         with StatusReporter(data, result, self._context, run) as status:
             if self._context.dry_run or not run:
                 self._run_iteration(data, result, run)
                 return
             if data.error:
                 raise DataError(data.error)
-            while self._should_run(data.condition, self._context.variables):
+            limit = WhileLimit.create(data.limit, self._context.variables)
+            while self._should_run(data.condition, limit, self._context.variables):
                 executed_once = True
                 try:
-                    self._run_iteration(data, result, run)
+                    with limit:
+                        self._run_iteration(data, result, run)
                 except BreakLoop:
                     break
                 except ContinueLoop:
                     continue
+                except WhileLimitExceeded:
+                    raise ExecutionFailed(
+                        "WHILE loop was aborted because it did not finish within the "
+                        f"limit of {limit}. Use the 'limit' argument to increase or "
+                        "remove the limit if needed."
```

Comment:
It seems to be common in Python to prefix all lines with `f` in cases like this.

Reply:
Could WhileLimit possibly raise ExecutionFailed directly? Catching WhileLimitExceeded just to raise another exception feels a bit odd.

Reply:
Yes it can. Have to get rid of the f-string then :(